core/src/main/java/com/linecorp/armeria/server/DecoratingServiceBindingBuilder.java

```diff
@@ -0,0 +1,199 @@
+/*
+ * Copyright 2019 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.server;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.List;
+import java.util.function.Function;
+
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.MediaType;
+import com.linecorp.armeria.server.RouteBindingBuilders.RouteBindingBuilder;
+
+/**
+ * A builder class for binding a {@code decorator} with {@link Route} fluently.
+ * This class can be instantiated through {@link ServerBuilder#decoratingRoute()}.
+ *
+ * <p>Call {@link #build(Function)} or {@link #build(DecoratingServiceFunction)}
+ * to build the {@code decorator} and return to the {@link ServerBuilder}.
+ *
+ * <pre>{@code
+ * ServerBuilder sb = new ServerBuilder();
+ *
+ * sb.decoratingRoute()                                // Configure a decorator with route.
+ *   .pathUnder("/api/users")
+ *   .build((delegate, ctx, req) -> {
```

Comment:
If I'm not mistaken, calling a method like `pathUnder` is required. It feels a bit weird to me that we have some mandatory parameters that are implicit, and others explicit in the `build` method. Actually it's thinking about this that made me wonder whether we need to expose this API at all in my other comment :)

Reply:
We have talked about reusing existing [`SeviceBindingBuilder`](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/ServiceBindingBuilder.java#L43) so `pathUnder` is not mandatory if we call a method which has a path like `get` or `post`, for example:
```java
sb.decoratingRoute()
  .get(".foo").post("bar").build(..)
```
I know what you are concerned and confused. We need to talk to improve user experience and the clearness.